BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/LoginActivity.kt

```diff
@@ -0,0 +1,194 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.biometricloginsample
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.text.Editable
+import android.text.TextWatcher
+import android.util.Log
+import android.view.inputmethod.EditorInfo
+import androidx.activity.viewModels
+import androidx.appcompat.app.AppCompatActivity
+import androidx.appcompat.widget.AppCompatEditText
+import androidx.biometric.BiometricManager
+import androidx.biometric.BiometricPrompt
+import androidx.lifecycle.Observer
+import com.example.biometricloginsample.databinding.ActivityLoginBinding
+
+
+/**
+ * 1) after entering "valid" username and password, login button becomes enabled
+ * 2) User clicks biometrics?
+ *   - a) if no template exists, then ask user to register template
+ *   - b) if template exists, ask user to confirm by entering username & password
+ */
+class LoginActivity : AppCompatActivity() {
+    val TAG = "LoginActivity"
+
+    private lateinit var biometricPrompt: BiometricPrompt
+    private lateinit var cryptographyManager: CryptographyManager
+    private lateinit var encryptedServerTokenWrapper: CiphertextWrapper
+    private lateinit var binding: ActivityLoginBinding
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityLoginBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+    }
+
+    override fun onResume() {
+        super.onResume()
+        cryptographyManager = CryptographyManager()
+        val ciphertextWrapper = cryptographyManager.getCiphertextWrapperFromSharedPrefs(
+            applicationContext,
+            SHARED_PREFS_FILENAME,
+            Context.MODE_PRIVATE,
+            CIPHERTEXT_WRAPPER
+        )
+
+        if (ciphertextWrapper != null) {
+            if (SampleAppUser.fakeToken == null) {
+                encryptedServerTokenWrapper = ciphertextWrapper
+                showBiometricPromptForDecryption()
+            } else {
+                // The user has already logged in, so proceed to the rest of the app
+                // this is a todo for you, the developer
+                updateApp(getString(R.string.already_signedin))
+            }
+        } else {
+            loginWithPasswordManager()
+        }
+
+        binding.useBiometrics.setOnClickListener {
+            if (ciphertextWrapper != null) {
+                encryptedServerTokenWrapper = ciphertextWrapper
+                showBiometricPromptForDecryption()
+            } else {
+                startActivity(Intent(this, EnableBiometricLoginActivity::class.java))
+            }
+        }
```

Comment:
All of this should probably be in `onCreate`.

Reply:
Just a matter of modular preference

Reply:
No, it's a matter of consistency within our sample code.

Reply:
Please take a moment to run the code in your IDE. If you move the logic from onResume to onCreate, you will introduce a bug in the flow.

Reply:
Can you elaborate, @isaidamier ?
In case this has to be in `onResume` it should be documented as well, so others don't run into issues here.

Reply:
Can you elaborate, @isaidamier ?
In case this has to be in `onResume` it should be documented as well, so others don't run into issues here.

Reply:
added notes in-code.